File: Medium/Merge k Sorted Arrays/merge-k-sorted-arrays.cpp
//{ Driver Code Starts
//Initial Template for C++

#include<bits/stdc++.h>
#define N 105
using namespace std;
void printArray(vector<int> arr, int size)
{
for (int i=0; i < size; i++)
	cout << arr[i] << " ";
}


// } Driver Code Ends


class Solution {
public:
    vector<int> merge(vector<int>& a1, vector<int>& a2) {
        int len1 = a1.size();
        int len2 = a2.size();
        int i = 0, j = 0;
        vector<int> merged;
        while (i < len1 && j < len2) {
            if (a1[i] < a2[j]) {
                merged.push_back(a1[i]);
                i++;
            } else {
                merged.push_back(a2[j]);
                j++;
            }
        }
        while (i < len1) {
            merged.push_back(a1[i]);
            i++;
        }
        while (j < len2) {
            merged.push_back(a2[j]);
            j++;
        }
        return merged;
    }
    vector<int> mergeKArraysHelper(vector<vector<int>>& arr, int start, int end) {
        if (start == end) {
            return arr[start];
        }
        int mid = (start + end) / 2;
        vector<int> arr1 = mergeKArraysHelper(arr, start, mid);
        vector<int> arr2 = mergeKArraysHelper(arr, mid + 1, end);
        return merge(arr1, arr2);
    }
    vector<int> mergeKArrays(vector<vector<int>>& arr, int K) {
        if (K == 0) return vector<int>();
        return mergeKArraysHelper(arr, 0, K - 1);
    }
};


//{ Driver Code Starts.

int main()
{
	int t;
	cin>>t;
	while(t--){
	    int k;
	    cin>>k;
	    vector<vector<int>> arr(k, vector<int> (k, 0));
	    for(int i=0; i<k; i++){
	        for(int j=0; j<k; j++)
	        {
	            cin>>arr[i][j];
	        }
	    }
	    Solution obj;
    	vector<int> output = obj.mergeKArrays(arr, k);
    	printArray(output, k*k);
    	cout<<endl;
    }
	return 0;
}






// } Driver Code Ends

File: Medium/Merge k Sorted Arrays/merge-k-sorted-arrays.py
class Solution: 
    def merge(self, a1, a2):
        len1 = len(a1)
        len2 = len(a2)
        merged = []
        i = 0
        j = 0
        while i < len1 and j < len2:
            if a1[i] < a2[j]:
                merged.append(a1[i])
                i += 1
            else:
                merged.append(a2[j])
                j += 1
        while i < len1:
            merged.append(a1[i])
            i += 1
        while j < len2:
            merged.append(a2[j])
            j += 1
        return merged
    def mergeKArrays(self, arr, K):
        if K == 1:
            return arr[0]
        mid = K // 2
        b1 = self.mergeKArrays(arr[:mid], mid)
        b2 = self.mergeKArrays(arr[mid:], K - mid)
        result = self.merge(b1, b2)
        return result

#{ 
 # Driver Code Starts
#Initial Template for Python 3

if __name__=="__main__":
    t=int(input())
    for _ in range(t):
        n=int(input())
        numbers=[[ 0 for _ in range(n) ] for _ in range(n) ]
        line=input().strip().split()
        for i in range(n):
            for j in range(n):
                numbers[i][j]=int(line[i*n+j])
        ob = Solution();
        merged_list=ob.mergeKArrays(numbers, n)
        for i in merged_list:
            print(i,end=' ')
        print()
# } Driver Code Ends